### PR TITLE
Make tensorflow backend fully mimic theano.dot

### PR DESCRIPTION
**Pull request by the-moliver**
This should make the Tensorflow backend fully mimic the broadcasting semantics of the theano.dot function. Previously it could only handle one specific case.

### CODE REVIEW

**Comment by the-moliver** [comment]
Seems not work with shape inference due to: ```# TODO: try to auto-infer shape if exception is raised by get_output_shape_for```. At least I think that's what is going on. It's not doing anything too crazy, just permuting dimensions and reshaping before and after a matrix multiply.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -108,18 +108,17 @@ def cast(x, dtype):
...
-    if ndim(x) == 2 and ndim(y) == 3:
-        slices = []
-        for i in range(int_shape(y)[0]):
-            slice_i = tf.matmul(x, y[i, :, :])
-            slice_i = expand_dims(slice_i, 1)
-            slices.append(slice_i)
-        out = tf.concat(1, slices)
-        return out
+    if ndim(x) > 2 or ndim(y) > 2:
+        x_shape = [i.__int__() for i in x.get_shape()]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: You can use `int_shape()` (backend function).

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -108,18 +108,17 @@ def cast(x, dtype):
...
+        y_shape = [i.__int__() for i in y.get_shape()]
+        y_permute_dim = list(range(ndim(y)))
+        y_permute_dim = [y_permute_dim.pop(-2)] + y_permute_dim
+        return tf.reshape(tf.matmul(tf.reshape(x, [-1, x_shape[-1]]), tf.reshape(tf.transpose(y, perm=y_permute_dim), 
+                                                        [y_shape[-2], -1])), x_shape[:-1]+y_shape[:-2]+y_shape[-1:])
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Line is not PEP8, please fix indent and insert spaces around operators.

Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Also I think the line does too much, breaking it into several lines would be preferable.

**Pull request closed without merging**